### DISCUSSION TITLE
The author is far from realizing the importance of a good IDE

### DISCUSSION THREAD

**Discussion by 32ns** [vlang/v#14868] [General]
I have been following V for more than a year, and now I don’t even have a good IDE. VLS reports various errors. I can’t rely on guessing for every function. The author has been thinking about the problem from the perspective of God, and feels that simple language does not require powerful editing. That's because you think of every function! Does it make sense to build a Vinix in the early stage? Only when the basic functions are perfect will more people be willing to join in, read the unclear documents, guess Not necessarily the right function, how can life waste so much time!

**Comment by 32ns** [comment]
I feel that V should first write an IDE that is easy to use, so that people who use it can quickly get started, attract more people to join, and then make C2v better, and teach others how to transfer the C project to v. after all, V has too few libraries. Only with the support of these two points can v be put into the production environment

**Comment by JalonSolov** [comment]
The English idiom is "putting the cart before the horse".  What good would a "fancy" IDE do if the language is not stable enough for it to be of any use?  That's what you would have if you created an IDE before the language was at least mostly finished.

In other words, I think things have gone the way they should - now that V is mostly stable, it is time to write an IDE to make it easier for new people to use.

**Comment by medvednikov** [comment]
Please report VLS issues via github (include the code you used), and they will be fixed

**Reply by 32ns** [reply]
VLS is basically unavailable, which is why I didn't submit the issue under VLS

**Reply by 32ns** [reply]
It is recommended that you make a paid conversion between C2v and go2v libraries, which can not only help other developers, but also find problems with C2v and go2v. You can also earn more money. Sending out the conversion process can also expand the tutorial library

**Reply by JalonSolov** [reply]
> VLS is basically unavailable, which is why I didn't submit the issue under VLS

Since when?  I have been using it for a long while, now.  It works with vscode and sublime text 4, that I have tried.  Both those editors have plugins for V that work with vls.

https://github.com/vlang/vls

**Reply by JalonSolov** [reply]
> It is recommended that you make a paid conversion between C2v and go2v libraries, which can not only help other developers, but also find problems with C2v and go2v. You can also earn more money. Sending out the conversion process can also expand the tutorial library

I have no idea what you mean, here.  c2v and go2v are tools, not libraries.  Both are available for use, and both are freeware, although go2v is still in early stages.  c2v is usable just with `v translate`.

**Reply by 32ns** [reply]
Google Translate doesn't quite understand what I mean, I just want to say that paying to help others convert C to V, such as his existing or using lib, there is no corresponding library in V, because of various problems and convert it by yourself No, the author can help him convert, and then publish the library and the conversion process, so that the documents and libraries can be enriched, and the author can earn some money

**Reply by medvednikov** [reply]
If you're not going to report VLS bugs, they won't get fixed.

**Reply by Ouri028** [reply]
> Google Translate doesn't quite understand what I mean, I just want to say that paying to help others convert C to V, such as his existing or using lib, there is no corresponding library in V, because of various problems and convert it by yourself No, the author can help him convert, and then publish the library and the conversion process, so that the documents and libraries can be enriched, and the author can earn some money

Why not just call C from V? I find that what V offers in terms of builtin features caters to 90% of what most people would normally use, also I feel that libraries aren't really necessary at the moment since V has so many features out of the box.

Also, do you really need to convert C libraries to V, I don't really use a lot of C, but recently I did need a Tiled map loader and just called a C library (cute_headers) from V which was really easy.

But that's just my opinion.

**Reply by JalonSolov** [reply]
The main reason to convert C to V is to be able to maintain the code in a more modern/"sane" environment.  Especially if you want to work with something older, that hasn't gotten updates in a while, and needs some.

For new programmers, it also means you only need to learn V, instead of both V and C.

**Comment by Wajinn** [comment]
I'm a little bit confused by the OP's post. In many cases, people can use an editor, of which there are some that are listed below (to include the versatile Notepad++, which is somehow not on the awesome list or being updated).  V is doing relatively well in this area, in terms of making itself easier to use with popular editors.

https://github.com/vlang/awesome-v#syntax-highlighting
https://github.com/vlang/awesome-v#editor-plugins
https://github.com/vlang/notepad-plus-plus (though its getting a bit old)

There is an early solo attempt at creating an IDE for V and purely written in V (https://github.com/IsaiahPatton/Vide/), by IsaiahPatton (who has some other great V projects as well).

However, I'm of the opinion that an IDE for V would be more beneficial in terms of building GUIs, or at least an official/semi-official form builder that can be easily utilized by various editors. This is also an area where a lot of other programming languages have overlooked to their detriment or realized far too late, sometimes never.  Thus, it seems to have slowed or becomes a stumbling block with their popularity among general users (particularly new ones).

**Comment by zhuxiujia** [comment]
i think support [intellij idea](https://www.jetbrains.com/idea/) or [clion](https://www.jetbrains.com/clion/) and debug.

This IDE is by far the best experience.

I don't understand the implementation of the debug function, is it compiled into C first and then this breakpoint?

**Comment by ghost** [comment]
I have used Delphi, VisualStudio, Eclipse, and IntelliJ, so I am familiar with the convenience of such integrated development environments.

I am currently writing V programs without VLS in VSCode.
I make use of various plug-ins of VSCode.
VLS still has a lot of problems and is difficult to use in practice.
I am not young and my memory is not good, so I strongly prefer to use the help of LS, but I am still comfortable enough to code.

Do you want an integrated development environment on the level of VisualStudio or IntelliJ?
Look at Eclipse, an open source IDE.
You can think how much effort has been put into it.
Which do you want first, for the V language itself to reach a level where it can be fully utilized, or for an integrated development environment on the level of Eclipse?
I want the language itself to be fully functional first.
Both together? That's not possible. It would slow down the development of the language. I don't want that.
The number of volunteer developers is small.

However, I understand that you would like to see more emphasis on VLS than go2v.
Alex is thinking of VLS important. It's just that it's not yet fully developed.
Please participate in the development of VLS by reporting bugs and contributing.

If you still want an IDE more immediately, then by all means, fund it yourself and hire a dedicated programmer to develop it for you.
Remember.
Both V and VLS are open source.
They are basically developed by free volunteers.
You want to use V for your business. That's great.
So what do you pay for it?
Bug reports? Contributions? Funds?
If you can't report bugs or contribute source code, you can fund it.

**Reply by zhuxiujia** [reply]
I often use Clion to develop Rust, and I have compared vscode to develop Rust, and I find Clion to be the best experience by comparison. For example, with macro support, vscode relies on compilation to produce results, while Clion uses its own analysis engine to scan code and produce results directly

**Reply by ghost** [reply]
Yes, I know. But JetBrains has 1,900 employees.
https://www.jetbrains.com/company/

**Reply by 32ns** [reply]
You like it, and you're willing to memorize all the functions, case, and arguments. But most people don't have the time or the energy. Languages are written for writing.Does it make sense for me to call a function and look at the source code?I just stand in the perspective of an ordinary person in the elaboration of the problem, just want to make Vlang better development, put forward my suggestions, just suggestions

**Reply by medvednikov** [reply]
No one asks you to memorize all functions.

VLS works. If it doesn't work for you, submit bug reports, they will be fixed.

**Comment by ghost** [comment]
I would guess that debugging is the biggest frustration for people who have only developed with IDE.
This problem will not be solved until native-backend is completed.
As a matter of fact, I too rely exclusively on `dump()` for debugging.
This is the fate of transpilers, so there is nothing we can do about it.
Is there any good solution in this regard?

**Reply by medvednikov** [reply]
This is not true. Debugging works just fine in V.

![image](https://user-images.githubusercontent.com/687996/177261990-e0397f71-89d9-4ab7-82e9-818f8d48900e.png)

**Reply by medvednikov** [reply]
![image](https://user-images.githubusercontent.com/687996/177262081-fd51fff8-03ea-4ec3-b839-044ccba84381.png)

**Reply by medvednikov** [reply]
> It's using the gdb plugin
> It works by doing a debug build of the V code and then running it with gdb using the debugger

**Reply by ghost** [reply]
I know that, but the variables are C variables, right?
I can't see the values as they are in the V struct, can I?
Or has that been made possible recently?

**Reply by ghost** [reply]
For example, in the last image `a: World`, I want to see the value of each field in the V's `string` struct.

**Reply by ghost** [reply]
For example, if you want to check the variable `a.len`, you have to `dump(a.len)`.

**Reply by ghost** [reply]
Such situations are frequent, and if I want to check the value change of a struct field, I have to `dump(MyStruct)`, but when this has a reference field, `dump()` is too deep, and printing prints a large amount of content.
https://github.com/vlang/v/issues/13370

**Reply by ghost** [reply]
Then, the fields other than the reference field must be listed.
```v
dump(mystruct.abc)
dump(mystruct.def)
dump(mystruct.ghi)
....
```
If these areas are improved, debugging will be much easier.

**Reply by zhuxiujia** [reply]
Java also provides its own interface design for debugging, which is Java Platform Debugger Architecture (JPDA). Every time we use the debug function, we rely on JPDA support implementation.

so.Where you want to interrupt, add an int3 instruction (and record the original value). The instruction code is 0xcc. When the target executes to INT3, it generates an interrupt.

I wonder if V should provide this functionality to support normal debug functionality?

Perhaps V provides interrupts and V provides memery variable value and structure of v in memory， will IDE support be simpler?

**Reply by i582** [reply]
Btw, JetBrains plugin can display complex types, even channels or maps in debugger without native backend:

![](https://media.discordapp.net/attachments/1036451192307458144/1057391015679512708/2022-12-28_00.14.04.png)

![](https://media.discordapp.net/attachments/1036451192307458144/1047202769364328510/2022-11-29_21.28.41.png)

![](https://media.discordapp.net/attachments/1036451192307458144/1058007271105380452/2022-12-29_17.02.44.png)

**Comment by ddkwork** [comment]
manjaro Linux debug test falid,Windows11 was ok.

**Comment by zhuxiujia** [comment]
maybe,  Programming languages that provide interrupt instructions and debugging information may be a solution

**Comment by shove70** [comment]
1. in just two to three years, V got 29.9k stars. What does it mean?

	V has distinctive characteristics. It is an integrator. rust, golang, dlang, nim... It has many characteristics and is easy to attract people's attention.

2. is the community active? Are there many people left? What does it say?

	(1). The integrity is not enough for large-scale production:

		1) The standard library is not perfect;
		2) Lack of coroutine, etc.

	(2). The core competitiveness is not obvious:

		Fast   -- V is fast, as fast as C
		Simple -- and simpler
		Safety -- V is not unique
		No GC  -- V is not unique
		Compilation speed -- V is fast, but not the fastest

3. how to make people pay attention to and actively participate in production for a long time?

	(1). integrity, including grammatical stability and the completion of the standard library.
	(2). Provide coroutine and parallels features.
	(3). Improve V's core competitiveness: rust, golang and dlang have no good GUI implementation. Although there are many official libraries or third-party libraries, few are really easy to use and stable. If V can provide a Delphi form like designer with drag and drop design according to the author's intention, it will be a great competitive advantage! After the GUI module of V is complete, how wonderful it is to develop ide through v.gui module.

4 about IDE

	(1). How about goland? Are there many users? Many people are still using vscode, because many people are unwilling to switch back and forth in multiple ides, so the vast majority of people are using vscode
	(2). IDE is not the most important at present. It is enough to make VLS more stable. When everything is complete, it will be icing on the cake to launch the ide of V.

**Reply by JalonSolov** [reply]
Take a look at https://github.com/nedpals/intellij-vlang - it should work with IDEA, GoLand, PHPStorm, or any of the other Intellij IDEs.  Of course, it's just started, but you can look at the page to see what's working so far.

**Comment by i582** [comment]
Please, see https://intellij-v.github.io

**Reply by 32ns** [reply]
The home page is beautifully done

**Comment by Wajinn** [comment]
Well, hopefully next to show up is a GUI/form builder, that could be used with the different editors/IDEs.

**Reply by i582** [reply]
We can start with online builder, I don't think that we need desktop app for first time

**Comment by bogen85** [comment]
A good programming language should not be favored by an IDE for it.

While an IDE does add value and is important, nothing in the language design should in any way depend or favor an IDE.
In other words, no feature or change to a language or tooling should make non-IDE usage less favorable than IDE usage.

While I may want to allow language or tooling changes to make non-IDE usage more favorable than more than IDE usage, I understand that would not be desirable overall.

So yes, I do recognize the value of a good IDE, but it should never be at the expense of non-IDE usage in regard to changes to the language or tooling.

**Reply by i582** [reply]
I think it's still not about the language features for the IDE, but for newcomers who come to the language from Go, Rust, Java, even C/C++ where there are first-class IDEs. 

Of course, you can debug with dumps, write without autocomplete, and go to the documentation site to find the right function every time until you remember them. 

However, why is it for an ordinary user? 

He will go and open IDEA/CLion/GoLand and drink coffee and enjoy some nice coding where you need to solve the problems of the task, and not remember the right function or scroll through a bunch of compiler errors that he will see only after compilation.

**Reply by bogen85** [reply]
There are many editors are not full blown IDEs that satisfy a lot of things (autocomplete and more, without being one of the heavy weight IDEs most think of when IDE is mentioned).

I agree that the heavy weight IDEs are extremely beneficial to many ordinary users.

**Comment by jkvatne** [comment]
I have installed the V-lang plug-in for Jetbrains CLion. It works very well. I can set breakpoints in the V code, and I can inspect variables the same way as in other languages. It also shows documentation and jumps to implementation of functions/variables. The normal Run/Debug button works as expected. The only thing missing is that hovering over a variable will not show its value. You must look at the "Variables" panel, where all variables are shown.  An other small issue is that the plugin does not support the latest CLion version yet. I must use version 2023.1.5. But the Jetbrains Toolbox makes it very easy to select version.

**Reply by JalonSolov** [reply]
Unfortunately, that plugin is no longer maintained (and can't be - he deleted the source).

We will have to wait until someone else provides a replacement.

**Reply by jkvatne** [reply]
But someone must have a copy....
It is too good just to let go.

**Reply by JalonSolov** [reply]
No.  There is no copy.  It is gone.

Others have been looking into a replacement, but nobody has mentioned any progress so far.

**Reply by spectatorBH** [reply]
What's the big drama behind that sudden move by the author of the plugin not just to drop any further development but even to fully kill the public repo? 

And what about the VASCO project death? Is this all interconnected? I don't get it: ppl start to do a lot of useful work for community, make good progress, and then all of a sudden drop it all silently and disappear without publicly saying a word. What's going on?

**Reply by JalonSolov** [reply]
There was "drama" in the discord channels.  It's done, now, but at this cost.

**Reply by Wajinn** [reply]
> What's the big drama behind that sudden move by the author of the plugin not just to drop any further development but even to fully kill the public repo? And what about the VASCO project death? 

The drama, concerning those parties, always existed and even before the plugin and VASCO. There always seemed to be this strangeness,  jabbing, and odd potshots being thrown. Myself and probably others were hopeful that something positive would come of it and despite the strangeness surrounding it, but not surprised at all with the direction it appears to have finally taken.

**Reply by ArtemkaKun** [reply]
Hi @spectatorBH , check the following links for more information:
- https://github.com/walkingdevel
- v-analyzer/v-analyzer@9ac84e6
- https://youtu.be/cUv13In9RP4?si=qMRQZ1BIzhLsERnY

**Reply by 32ns** [reply]
I have a copy, but I don't think it's very respectful of the author to publish it without the author's consent. He must not want it to be used again. I don't know what happened.

**Reply by jlmess77** [reply]
If it has a license like MIT, GPL or whatever on those files. You can release it without disrespect to the author unless you are a weak sauce then have a happy life.

**Reply by ylluminate** [reply]
@32ns Petr has given approval to use source, explained that it was (considered) MIT, see: https://discord.com/channels/592103645835821068/592106336838352923/1205187933993173033

> ... anyone can use the code as you want, if you have it...

So there is apparent community interest if you are able to share this.

**Reply by 32ns** [reply]
https://github.com/32ns/intellij-v

**Reply by jkvatne** [reply]
Thank you!

**Reply by medvednikov** [reply]
Thanks. Turns out I had an old clone from 2023-03-24.

I've added missing stuff from your repo since yours is the more recent one:

```
102 files changed, 707 insertions(+), 305 deletions(-)
```

I'll be using that because it preserves the git history.

https://github.com/vlang/intellij-v

**Reply by 32ns** [reply]
I'm not familiar with java, I hope my friends in the community can update it.

**Reply by Wajinn** [reply]
Well glad that it seems this is working out and that it was under MIT. Noticed that people were asking for it to be updated, and that appears to be the direction it is going. An updated release will likely make many happy.

**Comment by Androthi** [comment]
Years ago when I started programming in the high level assembly language, there was no IDE for it. so I wrote one.
I don't think it can be said that IDEs are in lacking supply these days. but if you're not happy with what's available, we'll, there's a project for you.

**Comment by medvednikov** [comment]
No worries, this year we'll release a big update to the VSCode plugin. It'll have autocomplete and debugging (something that the free version of intellij doesn't support).

**Reply by jkvatne** [reply]
Well, intellij has debugging when you use CLion, except that "step into" does not work. But having both VSCode and Intellij supporting V is a big pluss. Personally I like Intellij better, and I have the JetBrains "All products pack". I find it is worth the money.
